File: spare_parts_app_new/lib/models/product.dart
class Product {
  final int id;
  final String name;
  final String partNumber;
  final String? rackNumber;
  final double mrp;
  final double sellingPrice; // General/Default price
  final double wholesalerPrice;
  final double retailerPrice;
  final double mechanicPrice;
  final int stock;
  final int wholesalerId;
  final String? imagePath;
  final bool enabled;

  Product({
    required this.id,
    required this.name,
    required this.partNumber,
    this.rackNumber,
    required this.mrp,
    required this.sellingPrice,
    required this.wholesalerPrice,
    required this.retailerPrice,
    required this.mechanicPrice,
    required this.stock,
    required this.wholesalerId,
    this.imagePath,
    this.enabled = true,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'],
      name: json['name'],
      partNumber: json['partNumber'],
      rackNumber: json['rackNumber'],
      mrp: (json['mrp'] as num).toDouble(),
      sellingPrice: (json['sellingPrice'] as num).toDouble(),
      wholesalerPrice:
          (json['wholesalerPrice'] ?? json['sellingPrice'] as num).toDouble(),
      retailerPrice:
          (json['retailerPrice'] ?? json['sellingPrice'] as num).toDouble(),
      mechanicPrice:
          (json['mechanicPrice'] ?? json['sellingPrice'] as num).toDouble(),
      stock: json['stock'],
      wholesalerId: json['wholesalerId'] ?? 0,
      imagePath: json['imagePath'],
      enabled: (json['enabled'] ?? 1) == 1,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'partNumber': partNumber,
      'rackNumber': rackNumber,
      'mrp': mrp,
      'sellingPrice': sellingPrice,
      'wholesalerPrice': wholesalerPrice,
      'retailerPrice': retailerPrice,
      'mechanicPrice': mechanicPrice,
      'stock': stock,
      'wholesalerId': wholesalerId,
      'imagePath': imagePath,
      'enabled': enabled ? 1 : 0,
    };
  }

  Product copyWith({
    int? id,
    String? name,
    String? partNumber,
    String? rackNumber,
    double? mrp,
    double? sellingPrice,
    double? wholesalerPrice,
    double? retailerPrice,
    double? mechanicPrice,
    int? stock,
    int? wholesalerId,
    String? imagePath,
    bool? enabled,
  }) {
    return Product(
      id: id ?? this.id,
      name: name ?? this.name,
      partNumber: partNumber ?? this.partNumber,
      rackNumber: rackNumber ?? this.rackNumber,
      mrp: mrp ?? this.mrp,
      sellingPrice: sellingPrice ?? this.sellingPrice,
      wholesalerPrice: wholesalerPrice ?? this.wholesalerPrice,
      retailerPrice: retailerPrice ?? this.retailerPrice,
      mechanicPrice: mechanicPrice ?? this.mechanicPrice,
      stock: stock ?? this.stock,
      wholesalerId: wholesalerId ?? this.wholesalerId,
      imagePath: imagePath ?? this.imagePath,
      enabled: enabled ?? this.enabled,
    );
  }
}

File: spares-hub-server/src/main/java/com/spareshub/entity/Product.java
package com.spareshub.entity;

import jakarta.persistence.*;

@Entity
@Table(name = "products")
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;
    private String name;
    @Column(name = "part_number")
    private String partNumber;
    private Double mrp;
    @Column(name = "selling_price")
    private Double sellingPrice;
    @Column(name = "wholesaler_price")
    private Double wholesalerPrice;
    @Column(name = "retailer_price")
    private Double retailerPrice;
    @Column(name = "mechanic_price")
    private Double mechanicPrice;
    private Integer stock;
    @Column(name = "wholesaler_id")
    private Integer wholesalerId;
    @Column(name = "image_path")
    private String imagePath;

    public Integer getId() { return id; }
    public void setId(Integer id) { this.id = id; }
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    public String getPartNumber() { return partNumber; }
    public void setPartNumber(String partNumber) { this.partNumber = partNumber; }
    public Double getMrp() { return mrp; }
    public void setMrp(Double mrp) { this.mrp = mrp; }
    public Double getSellingPrice() { return sellingPrice; }
    public void setSellingPrice(Double sellingPrice) { this.sellingPrice = sellingPrice; }
    public Double getWholesalerPrice() { return wholesalerPrice; }
    public void setWholesalerPrice(Double wholesalerPrice) { this.wholesalerPrice = wholesalerPrice; }
    public Double getRetailerPrice() { return retailerPrice; }
    public void setRetailerPrice(Double retailerPrice) { this.retailerPrice = retailerPrice; }
    public Double getMechanicPrice() { return mechanicPrice; }
    public void setMechanicPrice(Double mechanicPrice) { this.mechanicPrice = mechanicPrice; }
    public Integer getStock() { return stock; }
    public void setStock(Integer stock) { this.stock = stock; }
    public Integer getWholesalerId() { return wholesalerId; }
    public void setWholesalerId(Integer wholesalerId) { this.wholesalerId = wholesalerId; }
    public String getImagePath() { return imagePath; }
    public void setImagePath(String imagePath) { this.imagePath = imagePath; }
}
